[DebugInfo][DWARF] Set is_stmt on first non-line-0 instruction in BB

SLTozer opened the pull request:
Fixes: https://github.com/llvm/llvm-project/issues/104695

This patch adds the is_stmt flag to line table entries for the first instruction with a non-0 line location in each basic block, to ensure that it will be used for stepping even if the last instruction in the previous basic block had the same line number; this is important for cases where the new BB is reachable from BBs other than the preceding block.

SLTozer committed:
[DebugInfo][DWARF] Set is_stmt on first non-line-0 instruction in BB

Fixes: https://github.com/llvm/llvm-project/issues/104695

This patch adds the is_stmt flag to line table entries for the first
instruction with a non-0 line location in each basic block, to ensure
that it will be used for stepping even if the last instruction in the
previous basic block had the same line number; this is important for cases
where the new BB is reachable from BBs other than the preceding block.

SLTozer committed:
Update existing tests with newly generated is_stmt flags

SLTozer commented:
This does result in some is_stmts being added that are unnecessary, i.e. cases where the last instruction in the current BB's sole predecessor has the same line number as the first instruction in the current BB. We could potentially include that in the analysis, but I think there's no way to do so without doing a more complex (though probably not too expensive) analysis in DwarfDebug; we also wouldn't be able to account for indirect branches.

I think in any case this patch is an improvement if we're considering such cases, as not only might the current BB not have the previous BB as its sole predecessor, but the previous BB might not even be a predecessor - making the is_stmt-elision completely incorrect.

pogo59 reviewed (commented):
I like this. I should have considered BB transitions when I did the line-0 accommodation in the first place. See question about the new test.

pogo59 commented on llvm/test/DebugInfo/Generic/is_stmt-at-block-start.ll:
```diff
@@ -0,0 +1,38 @@
+; RUN: %llc_dwarf -O0 -filetype=obj < %s | llvm-dwarfdump --debug-line - | FileCheck %s
+
+; CHECK:      {{0x[0-9a-f]+}}     13      5 {{.+}} is_stmt
+; CHECK-NEXT: {{0x[0-9a-f]+}}     13     25 {{.+}} is_stmt
```
ISTM the first check is for block if.then2, the second is for if.else4? Should there also be one for if.end8?

SLTozer replied:
Not for this test specifically, as only these two instructions have non-0 line numbers (excluding 0x0, which has the function header).

pogo59 replied:
Oh right, `!dbg` isn't inherited the way it is for asm.
But, you could use the same `!dbg` for both, showing that it really is the BB that triggers it.

SLTozer replied:
Hm, trying to make this work has revealed another an issue - when the lines are actually identical, they will be merged into a single entry (whether this patch is in effect or not), which any jump to the branch would land in the middle of. I'm not sure whether that would result in a step in the debugger or not; it prevents the test from working, at the very least.

pogo59 reviewed (approved):
LGTM with a test suggestion.

jryans reviewed (approved):
Looks like a very sensible improvement, thanks for working on this! 😄

SLTozer committed:
Update to not merge identical DLs across BB boundaries

SLTozer commented:
Had a free moment so updated this - the previous version (as noted in my response to @pogo59 above) would still merge adjacent DLs across BB boundaries if they were exactly identical, rather than differing in column number; I've changed this to now only merge adjacent DLs within a BB, which strictly reduces the cases where we'll merge identical DLs.

jryans reviewed (approved):
Seems like a good tweak, looks good overall to me.

pogo59 reviewed (commented):
Smooth. Still LGTM.

SLTozer committed:
Update other tests for new line output

SLTozer commented:
Updated a couple more tests whose outputs have changed as of the latest functional change - all of these look "correct" to me, in that they're adding is_stmt to instructions at the start of a basic block, or in `empty-inline.mir`, adding a few new entries that prevent a line-0 instruction at the start of a basic block from having a non-0 fallthrough location, and adding a new is_stmt entry at the start of a block (similar to the case in the initial report).

SLTozer commented:
I added the test to the `Generic/` test dir since it doesn't test target-specific behaviour, but it does rely on the basic block layout which varies between targets; added commit 616f7d3d4f6d9bea6f776e357c938847e522a681 to move the test into `X86/`, which should fix the buildbot errors.

SLTozer commented:
Hm, another [buildbot error](https://lab.llvm.org/buildbot/#/builders/160/builds/4142) - this time `createMCCodeEmitter failed`, which seems like it will also be a trivial fix, but I'll revert for now.

dwblaikie commented:
We're seeing a pretty significant size regression from this change in self-host builds of clang and some other internal binaries we track.

For an internal debug build (with Split DWARF, compressed debug info sections in object files but uncompressed in the linked binary) we saw a total 3.6% increase in the size of the clang binary, a 12.7% increase in the size of the .debug_line section in object files, and a 18.3% increase in the size of .debug_line in the linked executable.

Might be worth reverting this and reconsidering whether the growth is acceptable or whether there's a more targeted fix that can be made?

SLTozer commented:
> Might be worth reverting this and reconsidering whether the growth is acceptable or whether there's a more targeted fix that can be made?

I have another patch up now that tries to be more cautious with its application of is_stmt: https://github.com/llvm/llvm-project/pull/108251

It currently needs a fix, and confirmation that it has suitable performance, but it should result in fewer is_stmts being applied. With that said, if it doesn't bring the numbers down that much, there is a reasonable question about correctness vs size - it might not be feasible to compute the minimum set of is_stmts that don't result in missing steps, so there might be either a tradeoff to make, or a flag to add.

dwblaikie commented:
Could you put this behind a flag or revert this for now (to see if you can find a solution that's sufficiently cautious to lower the size cost to something more acceptable) - I think the growth is a bit big to take by default? (perhaps other people have other opinions, though, in which case I'm open to that/we can take the hit if that's the prevailing opinion)

SLTozer commented:
> Could you put this behind a flag or revert this for now

I can revert it, at least until the next patch is ready to land. I've tried to investigate this locally to see if the situation is improved by the next patch along, but my local build isn't showing anywhere close to the same increase - I've got .debug_line size increasing by 1.3% as a result of this commit. When you say "Debug", does that mean the CMake "Debug" option, or a release build with debug info enabled?
